### ISSUE TITLE
Question: Serializing records with sets and lists in F#

### ISSUE THREAD

**Issue by bhandfast** [msgpack/msgpack-cli#178]
I'm using F# and I have trouble packing record types containing sets or lists.

The code below fails with the message:

> "System.Runtime.Serialization.SerializationException: Type 'FSI_0003+SetRecord' does not have default (parameterless) public constructor."

I guess I need to write a new custom serializer that packs the values as arrays are packed but I'm not sure how? Any ideas or pointers to where I can find examples?

``` fsharp
#r "MsgPack.dll"

type TestRecord = {
    Foo : string;
    Bar: int;
}

type OptionRecord = {
    Foo : string option;
    Bar: int;
}

let serializer = MsgPack.Serialization.SerializationContext.Default

let serializeToMsgPack<'T> (x:'T) = 
    serializer.GetSerializer<'T>().PackSingleObject(x)

// Gives result:

serializeToMsgPack { TestRecord.Foo = "Hello"; Bar = 42 }
serializeToMsgPack { OptionRecord.Foo = Some "Hello"; Bar = 42 }
serializeToMsgPack { OptionRecord.Foo = None; Bar = 42 }

type SetRecord = {
    Foo : string option;
    Bar: Set<int>;
}


// Crashes:
serializeToMsgPack { SetRecord.Foo = Some "Hello"; Bar = Set.empty.Add(42) }


type ListRecord = {
    Foo : string option;
    Bar: int list;
}

// Crashes:

serializeToMsgPack { ListRecord.Foo = Some "Hello"; Bar = [ 1 ] }
```

**Comment by rodrigovidal** [comment]
Hey @bhandfast, try to add the [<CLIMutable>] attribute to those records.

**Comment by bhandfast** [comment]
Thanks for the help. However, now I get another error.

> System.Runtime.Serialization.SerializationException: Type 'Microsoft.FSharp.Collections.FSharpSet`1[System.Int32]' does not have both of default (parameterless) public constructor and  public constructor with an Int32 parameter.
   at MsgPack.Serialization.UnpackHelpers.GetCollectionConstructor(Type instanceType)
   at MsgPack.Serialization.AbstractSerializers.SerializerBuilder`2.BuildCollectionCreateInstance(TContext

 context, Type concreteType)

Arrays seems fine though, guess I could remodel my type.

**Comment by yfakariya** [comment]
Sorry for delay. There are 2 problems.
1. There is a bug that record like type cannot be serialized if it has mutable collection type member.
1. As you know, `FSharpSet<T>` is not supported. You must register custom serializer for `FSharpList`, `FSharpSet`, and `FSharpMap` as built-in immutable collections support.

**Comment by yfakariya** [comment]
Fixed in 7bfdee5 for the bug. F# collection will not be supported in built-in serializers.
